torch/autograd/__init__.py

```diff
@@ -26,32 +26,47 @@
 __all__ = ['Variable', 'Function', 'backward', 'grad_mode']
 
 _OptionalTensor = Optional[torch.Tensor]
+_NtSize = Union[Tuple[torch.Tensor], List[torch.Tensor]]
+_ShapeorNestedShape = Union[_NtSize, _size]
+
+def _calculate_shape(output: torch.Tensor, grad: torch.Tensor,
+                     is_grads_batched: bool) -> Tuple[_ShapeorNestedShape, _ShapeorNestedShape]:
+    if output.is_nested:
+        if is_grads_batched:
+            raise RuntimeError("Batched grads are not supported with Nested Tensor.")
+        grad_shape = grad.nested_size()
+        out_shape = output.nested_size()
+        return out_shape, grad_shape
+
+    reg_out_shape = output.shape
+    reg_grad_shape = grad.shape if not is_grads_batched else grad.shape[1:]
+    return reg_out_shape, reg_grad_shape
 
 def _make_grads(outputs: Sequence[torch.Tensor], grads: Sequence[_OptionalTensor],
                 is_grads_batched: bool) -> Tuple[_OptionalTensor, ...]:
     new_grads: List[_OptionalTensor] = []
     for out, grad in zip(outputs, grads):
         if isinstance(grad, torch.Tensor):
-            grad_shape = grad.shape if not is_grads_batched else grad.shape[1:]
-            if not out.shape == grad_shape:
+            if not torch.has_same_shape(out, grad):
```

Comment:
doesn't handle batched_shape check write can change torch.has_same_shape(out, grad) to 
```
unbatched_grad = grad if not is_grads_batched else grad[0]
 torch.has_same_shape(out, unbatched_grad)
 ```
Need to test that with batched grads though